File: scripts/verify-predicates.ts
/**
 * Smoke test for the predicate ontology loader.
 *
 * Run: npx tsx scripts/verify-predicates.ts
 * Exit 0 on success, 1 on any failure.
 *
 * Checks mirror backend/scripts/verify_predicates.py so a drift between
 * the two loaders would show up as one passing and the other failing.
 */
import {
    PREDICATE_SPECS,
    PREDICATE_NAMES,
    ALLOWED_PREDICATES,
    ASYMMETRIC_PREDICATES,
    DAG_RELATIONS,
    PREDICATE_RUBRIC,
} from '../src/config/predicates';

const REQUIRED = new Set([
    'IS_A', 'PART_OF', 'REQUIRES', 'PRECEDES', 'EXTENSION_OF',
    'CAUSES', 'LED_TO', 'FOUND_IN', 'LOCATED_IN',
    'CONTRASTS_WITH', 'RELATES_TO',
]);

const SNAKE_CASE = /^[A-Z][A-Z0-9_]*$/;

const errors: string[] = [];
const check = (cond: boolean, msg: string) => {
    if (!cond) errors.push(msg);
};

// 1 + 2. Sanity floor
check(PREDICATE_SPECS.length >= 20,
    `Only ${PREDICATE_SPECS.length} predicates loaded (expected >= 20)`);

// 3. Required fields
PREDICATE_SPECS.forEach((p, i) => {
    check(!!p.name, `predicate[${i}]: empty name`);
    check(typeof p.description === 'string', `${p.name}: description not string`);
    check(typeof p.example === 'string', `${p.name}: example not string`);
    check(typeof p.asymmetric === 'boolean', `${p.name}: asymmetric not bool`);
    check(typeof p.dag === 'boolean', `${p.name}: dag not bool`);
});

// 4. No duplicates
const seen = new Set<string>();
const dupes = new Set<string>();
for (const n of PREDICATE_NAMES) {
    if (seen.has(n)) dupes.add(n);
    seen.add(n);
}
check(dupes.size === 0, `Duplicate predicates: ${[...dupes].sort().join(', ')}`);

// 5 + 6. Name shape
for (const n of PREDICATE_NAMES) {
    check(!n.includes(' '), `'${n}' contains whitespace`);
    check(SNAKE_CASE.test(n), `'${n}' is not UPPER_SNAKE_CASE`);
}

// 7. Subsets
const strayAsym = [...ASYMMETRIC_PREDICATES].filter((n) => !ALLOWED_PREDICATES.has(n));
const strayDag = [...DAG_RELATIONS].filter((n) => !ALLOWED_PREDICATES.has(n));
check(strayAsym.length === 0, `ASYMMETRIC not in ALLOWED: ${strayAsym}`);
check(strayDag.length === 0, `DAG not in ALLOWED: ${strayDag}`);

// 8. DAG => asymmetric
const dagNotAsym = [...DAG_RELATIONS].filter((n) => !ASYMMETRIC_PREDICATES.has(n));
check(dagNotAsym.length === 0,
    `DAG predicates must be asymmetric: ${dagNotAsym}`);

// 9. Rubric renders
check(PREDICATE_RUBRIC.length > 0, 'PREDICATE_RUBRIC is empty');
for (const n of PREDICATE_NAMES) {
    check(PREDICATE_RUBRIC.includes(n), `'${n}' missing from rubric`);
}

// 10. Regression guard
const missing = [...REQUIRED].filter((n) => !ALLOWED_PREDICATES.has(n));
check(missing.length === 0, `Required predicates missing: ${missing}`);

// Report
if (errors.length > 0) {
    console.log(`[verify-predicates] FAIL — ${errors.length} issue(s):`);
    for (const e of errors) console.log(`  ✘ ${e}`);
    process.exit(1);
}

console.log(
    `[verify-predicates] OK — ${PREDICATE_SPECS.length} predicates ` +
    `(${ASYMMETRIC_PREDICATES.size} asymmetric, ${DAG_RELATIONS.size} DAG)`,
);

File: backend/scripts/verify_predicates.py
"""
Smoke test for the predicate ontology loader.

Run: python -m backend.scripts.verify_predicates
     (or: cd backend && python -m scripts.verify_predicates)
Exit 0 on success, 1 on any failure.
"""
import re
import sys

try:
    from backend.graph.predicates import (
        PREDICATE_SPECS,
        PREDICATE_NAMES,
        ALLOWED_PREDICATES,
        ASYMMETRIC_PREDICATES,
        DAG_RELATIONS,
        predicate_rubric,
    )
except ImportError:
    from pathlib import Path
    sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
    from graph.predicates import (  # type: ignore
        PREDICATE_SPECS,
        PREDICATE_NAMES,
        ALLOWED_PREDICATES,
        ASYMMETRIC_PREDICATES,
        DAG_RELATIONS,
        predicate_rubric,
    )


REQUIRED = {
    "IS_A", "PART_OF", "REQUIRES", "PRECEDES", "EXTENSION_OF",
    "CAUSES", "LED_TO", "FOUND_IN", "LOCATED_IN",
    "CONTRASTS_WITH", "RELATES_TO",
}

SNAKE_CASE = re.compile(r"^[A-Z][A-Z0-9_]*$")


def _check(cond: bool, msg: str, errors: list[str]) -> None:
    if not cond:
        errors.append(msg)


def main() -> int:
    errors: list[str] = []

    _check(len(PREDICATE_SPECS) >= 20,
           f"Only {len(PREDICATE_SPECS)} predicates loaded (expected >= 20)", errors)

    for i, p in enumerate(PREDICATE_SPECS):
        _check(bool(p.name), f"predicate[{i}]: empty name", errors)
        _check(isinstance(p.description, str),
               f"{p.name}: description not a string", errors)
        _check(isinstance(p.example, str),
               f"{p.name}: example not a string", errors)
        _check(isinstance(p.asymmetric, bool),
               f"{p.name}: asymmetric not a bool", errors)
        _check(isinstance(p.dag, bool),
               f"{p.name}: dag not a bool", errors)

    dupes = [n for n in PREDICATE_NAMES if PREDICATE_NAMES.count(n) > 1]
    _check(not dupes, f"Duplicate predicates: {sorted(set(dupes))}", errors)

    for name in PREDICATE_NAMES:
        _check(" " not in name, f"'{name}' contains whitespace", errors)
        _check(bool(SNAKE_CASE.match(name)),
               f"'{name}' is not UPPER_SNAKE_CASE", errors)

    stray_asym = ASYMMETRIC_PREDICATES - ALLOWED_PREDICATES
    stray_dag = DAG_RELATIONS - ALLOWED_PREDICATES
    _check(not stray_asym, f"ASYMMETRIC not in ALLOWED: {stray_asym}", errors)
    _check(not stray_dag, f"DAG not in ALLOWED: {stray_dag}", errors)

    dag_not_asym = DAG_RELATIONS - ASYMMETRIC_PREDICATES
    _check(not dag_not_asym,
           f"DAG predicates must be asymmetric: {dag_not_asym}", errors)

    try:
        rubric = predicate_rubric()
        _check(len(rubric) > 0, "predicate_rubric() returned empty", errors)
        for name in PREDICATE_NAMES:
            _check(name in rubric, f"'{name}' missing from rubric", errors)
    except Exception as e:
        errors.append(f"predicate_rubric() raised: {e}")

    missing = REQUIRED - ALLOWED_PREDICATES
    _check(not missing, f"Required predicates missing: {missing}", errors)

    if errors:
        print(f"[verify_predicates] FAIL - {len(errors)} issue(s):")
        for e in errors:
            print(f"  X {e}")
        return 1

    print(f"[verify_predicates] OK - {len(PREDICATE_SPECS)} predicates "
          f"({len(ASYMMETRIC_PREDICATES)} asymmetric, {len(DAG_RELATIONS)} DAG)")
    return 0


if __name__ == "__main__":
    sys.exit(main())